Changing the base load balancer class over to use priorities

alyssawilk opened the pull request:
Changing the LoadBalancerBase to use priorities, following the simplistic strategy docced up in the eds proto, where we route to the highest priority level with non-zero healthy endpoints.

This results in simple priority-based load balancing for RoundRobinLoadBalancer, LeastRequestLoadBalancer, and RandomLoadBalancer.

*Risk Level*: Medium 
It's a big change but then again no one is using it :-)

*Testing*:
Existing tests parameterized to target P=0 and P=1 to verify equivalence.
Additional unit tests and one integration test explicitly testing primary/failover/... inteeractions

*Docs Changes*:
https://github.com/envoyproxy/data-plane-api/pull/326

*Release Notes*:
Added support for priorities for several types of load balancer (RoundRobinLoadBalancer, LeastRequestLoadBalancer, RandomLoadBalancer)

Continued progress on #1929

alyssawilk commented:
nevermind on the "should we do something non-naive yet?"  I think I know where I want to go with this and it should definitely be a series of PRs. I'll stick with the simple-but-stupid for the refactor

mattklein123 commented:
> nevermind on the "should we do something non-naive yet?" I think I know where I want to go with this and it should definitely be a series of PRs. I'll stick with the simple-but-stupid for the refactor

I talked to @alyssawilk offline and I agree that we should do the simple/stupid thing first and then make it better. (We almost definitely don't want this algorithm to be the final solution). But I think having this as close to possible to what we already do is probably a good place to start. Also, for others, we also discussed the fact that when I rejigger things to allow central/TLS load balancers, we can do some of these more expensive computations once and then fan out to all threads which will be more efficient in CPU and memory usage.

@alyssawilk in general this approach LGTM. LMK when finished and I can look in more detail.

mattklein123 commented:
@alyssawilk I don't have much time today, but I wanted to briefly mention that it occurred to me that we may have a problem with the "local host set." The whole idea behind the "local host set" is that we know how many senders there are so that we can balance traffic correctly over all of the target zones. In this sense, if we assume that usually all of the "local host set" are going to failover at roughly the same time, we don't care about the priority of the "local host set", we actually want to know *all* the hosts in the entire local host cluster so that we can do the calculations against the target priority groups, such that when we failover we balance properly against the failover group.

Hopefully ^ makes sense. Let me know if not. (It's possible that you already did this also - I didn't look at the algorithm changes very carefully yet).

alyssawilk commented:
per discussions with matt we're just going to enforce that local host sets don't have non-zero priorities.
Will reopen when that's done.

alyssawilk commented:
OK, reworked to not bother with multiple priorities for local host, added a TODO for config validation (let me know if you want it in this patch) and added more tests.  Let me know what you think!

htuch commented on RAW_RELEASE_NOTES.md:
```diff
@@ -12,3 +12,16 @@ final version.
+
+=============================================================
+
+Added supporting the priority field in eds LocalityLbEndpoints for several types of
```
Nit: s/eds/EDS, add Sphinx RST links to LocalityLbEndpoints optionally. Also, how come the ======== formatting vs. the normal bullet points that we've traditionally been using?

alyssawilk replied:
Simplified per the one line * format.  Should be good for another look at your leisure

htuch commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -12,6 +12,21 @@
+      return host_set.get();
+    }
+  }
+  return priority_set->hostSetsPerPriority()[0].get();
```
Can `priority_set->hostSetsPerPriority().size() == 0` here ever?

alyssawilk replied:
No, we always populate the first priority set.

mattklein123 replied:
+1 adding assert/comment here about this.

mattklein123 replied:
Still need assert/comment here I think?

htuch commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -44,18 +44,20 @@ class LoadBalancerBase {
   enum class LocalityRoutingState { NoLocalityRouting, LocalityDirect, LocalityResidual };
 
+  /**
+   * Increase per_priority_state_ to at leat |size|
```
Nit: s/leat/least/. Also, remove blank line. I think it would be helpful to state the invariant that per_priority_state_ must obey, i.e. the motivation for resizing above.

htuch commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -74,13 +76,33 @@ class LoadBalancerBase {
+  struct PerPriorityState {
+    uint64_t local_percent_to_route_{};
+    LocalityRoutingState locality_routing_state_{LocalityRoutingState::NoLocalityRouting};
+    std::vector<uint64_t> residual_capacity_;
```
I know this was an issue in the original, but I think it would help readability to explain what these fields purpose in life is.

htuch commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -74,13 +76,33 @@ class LoadBalancerBase {
-  Common::CallbackHandle* local_host_set_member_update_cb_handle_{};
+  void regenerateLocalityRoutingStructures(uint32_t priority);
+
+  uint32_t best_available_priority() { return best_available_host_set_->priority(); }
```
Nit, this and other methods might be `cont` candidates (for their side effects, not return).

htuch commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -92,42 +123,56 @@ void LoadBalancerBase::regenerateLocalityRoutingStructures() {
+void LoadBalancerBase::resizePerPriorityState(uint32_t size) {
+  size = std::max<uint32_t>(size, priority_set_.hostSetsPerPriority().size());
+  if (per_priority_state_.size() < size) {
+    for (size_t i = 0; i < size - per_priority_state_.size() + 1; ++i) {
```
If the target `size` is 2 and `per_priority_state_.size()` is 1, then I end up growing `per_priority_state_.size()` by 2 here; or is it late at night and I'm not counting correctly?

htuch commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -20,61 +35,77 @@ static const std::string RuntimePanicThreshold = "upstream.healthy_panic_thresho
+    // Update the host set to use for picking, based on the new state.
+    best_available_host_set_ = bestAvailable(&priority_set_);
+    // If there's a local priority set, regenerate all routing. If not, make
+    // sure per_priority_state_ is still large enough to not cause problems.
```
Can you elaborate in this comment on why we need to only regenerate all routing in this situation?

htuch commented on test/common/upstream/load_balancer_impl_test.cc:
```diff
@@ -48,55 +53,153 @@ class RoundRobinLoadBalancerTest : public LoadBalancerTestBase {
+  tertiary_host_set_.runCallbacks(add_hosts, {});
+  EXPECT_EQ(tertiary_host_set_.hosts_[0], lb_->chooseHost(nullptr));
+
+  // Now add a healthy host in P=0 and make sure it is immediately selected.
```
Is there similar coverage for when a previously health host is later marked unhealthy?

htuch commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -20,61 +35,83 @@ static const std::string RuntimePanicThreshold = "upstream.healthy_panic_thresho
+  });
+  if (local_priority_set_) {
+    // Multiple priorities are unsupported for local priority sets.
+    // TODO(alyssawilk) find the right place to reject config with this.
```
If multiple priorities aren't supported for the local priority set, i.e. it is always a singleton, could it just be a `HostSet` instead of a `PrioritySet`? I'm guessing there is some advantage in being able to treat it as a `PrioritySet` when working with common code, but just wanted to check (and maybe request a comment be added in some place to explain).

alyssawilk replied:
We could absolutely have local_host_set here.  I hacked that up a ways back and decided it wasn't much better because
- we have to convert from priority set to host set somewhere (the thread local cluster is just another cluster in the ClusterManagerImpl) and it's about as ugly elsewhere and 
- if we do it in ClusterManagerImpl we have to reinstate "add callback" to the upstream HostSet API (which makes me a bit nervous as generally it's the wrong thing to do to subscribe to updates to a host set).

Alternately [I think] we can make the "local envoy hosts" a host set in ClusterManagerImpl, but then we've moved the problem to config updates, as 99% would go to thread_local_clusters_ but updates to local_cluster_name.value() would be applied directly to the special one-off host set.

I don't have super strong feelings either way so take a look at where we set local_priority_set_ and let me know if you think one of those options is worth the extra code churn.

mattklein123 replied:
Per offline convo, if we were to actually support multiple priority levels for the local cluster, we would actually need to use *all priorities" when computing the locality routing structures if we assume that all Envoys in the local cluster failover at roughly the same time. To support this, IMO we would likely add additional methods at a PrioritySet to get all hosts or something like that. As such I think what we are doing here is probably good for now. I might add some more comments on what the "real" solution would be if we want to implement that ever. I might also add a TODO/comment about the "real" solution so future readers know all the thinking that went into this.

Also, per comment below, I do have a preference for blocking this at the EDS update layer and just using asserts in the LB code.

mattklein123 commented:
@alyssawilk I will give this a thorough review tomorrow.

mattklein123 commented on source/common/upstream/subset_lb.cc:
```diff
@@ -421,9 +421,9 @@ SubsetLoadBalancer::PrioritySubsetImpl::PrioritySubsetImpl(const PrioritySet& or
 }
 
-HostSetPtr SubsetLoadBalancer::PrioritySubsetImpl::createHostSet(uint32_t priority) {
+HostSetImpl* SubsetLoadBalancer::PrioritySubsetImpl::createHostSet(uint32_t priority) {
```
What's the reasoning behind changing this to return a raw pointer vs. unique_ptr with move semantics?

alyssawilk replied:
It allows calling host_subset->update() in the subset load balancer without down casts, because the subclass can override the return type.  Not doable if you encapsulate it in the unique pointer.  
If you prefer the casts and explicit ownership I'll switch back.

mattklein123 replied:
Sorry which is the code that breaks with unique_ptr? In general definitely prefer explicit ownership but curious what is breaking.

alyssawilk replied:
Or on the other hand it was for some reason which is since factored out since it works now.  Reinstated :-)

mattklein123 commented on source/common/upstream/ring_hash_lb.cc:
```diff
@@ -16,8 +16,13 @@ RingHashLoadBalancer::RingHashLoadBalancer(
+  priority_set.addMemberUpdateCb([this](uint32_t priority, const std::vector<HostSharedPtr>&,
+                                        const std::vector<HostSharedPtr>&) -> void {
+    if (priority != 0) {
+      return; // TODO(alyssawilk) make ring hash LB priority-aware.
```
Can we just block this inside EdsClusterImpl? The cluster knows it's LB type and we could check it there and fail the config. I'm not sure when/if we will make ring hash priority aware. Then we could switch this to assert.

In general, I'm wondering if we could do all the config blocking inside EdsClusterImpl. We could even add a ClusterManager method to get the "local cluster" so that an EDS cluster could tell if it is the local cluster and then fail if that cluster has more than 1 priority.

alyssawilk replied:
FWIW ring hash code is written and tested.  It depends on this one so I can't send it out yet. :-)

I'll take a look at what it'd take to do config validation here.  I'd wanted to do an end to end failed update in the load balancing test and I think that's going to be complicated.

mattklein123 replied:
If you are going to do it soon it's not a big deal, was just thinking in general it would be better to block config in EDS updates vs. silently ignore here.

alyssawilk replied:
Sure thing, done.

mattklein123 commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -74,13 +83,38 @@ class LoadBalancerBase {
+  // zero-priority host set if all host sets are fully unhealthy.
+  const HostSet* best_available_host_set_;
+
+  // The set of local Envoy instances which are load balancing across priority_set_
```
nit: period end of sentence.

mattklein123 commented on source/common/upstream/load_balancer_impl.h:
```diff
@@ -44,18 +44,27 @@ class LoadBalancerBase {
+  };
+
+  /**
+   * Increase per_priority_state_ to at least priority_set;.hostSetsPerPriority().size()
```
";." typo?

mattklein123 commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -20,61 +35,83 @@ static const std::string RuntimePanicThreshold = "upstream.healthy_panic_thresho
+  ASSERT(local_priority_set_);
   stats_.lb_recalculate_zone_structures_.inc();
+  // This should never happen unless someone uses unsupported priorities for local priority sets.
+  // Do error handling until such config is rejected, then it will change to an ASSERT
```
Per other comments can we just do this in this PR potentially?

mattklein123 replied:
Can this be switched to ASSERT now?

mattklein123 commented on source/common/upstream/load_balancer_impl.cc:
```diff
@@ -92,42 +129,54 @@ void LoadBalancerBase::regenerateLocalityRoutingStructures() {
-bool LoadBalancerBase::earlyExitNonLocalityRouting() {
-  if (host_set_.healthyHostsPerLocality().size() < 2) {
+void LoadBalancerBase::resizePerPriorityState() {
+  uint32_t size = priority_set_.hostSetsPerPriority().size();
```
nit: const

mattklein123 commented on RAW_RELEASE_NOTES.md:
```diff
@@ -11,6 +11,6 @@ final version.
 * Added transport socket interface to allow custom implementation of transport socket. A transport socket
   provides read and write logic with buffer encryption and decryption. The exising TLS implementation is
   refactored with the interface.
-* Added support for dynamic response header values (`%CLIENT_IP%` and `%PROTOCOL%`).
```
bad merge?

mattklein123 commented on source/common/upstream/cluster_manager_impl.h:
```diff
@@ -265,6 +266,8 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
   ClusterManagerInitHelper init_helper_;
   Config::GrpcMuxPtr ads_mux_;
   LoadStatsReporterPtr load_stats_reporter_;
+  // The name of the local cluster of Envoy instances if defined, else the empty string.
```
nit: "of this Envoy instance"

mattklein123 commented on source/common/upstream/eds.h:
```diff
@@ -41,6 +41,7 @@ class EdsClusterImpl : public BaseDynamicClusterImpl,
   std::unique_ptr<Config::Subscription<envoy::api::v2::ClusterLoadAssignment>> subscription_;
   const LocalInfo::LocalInfo& local_info_;
   const std::string cluster_name_;
+  const std::string local_cluster_name_;
```
nit: Would it be better to store a const reference to the ClusterManager here and have the localCluster() method be const on that? Not a big deal but thought I would mention it.

mattklein123 commented:
Needs master merge I think.

htuch commented on include/envoy/upstream/cluster_manager.h:
```diff
@@ -134,6 +134,13 @@ class ClusterManager {
+   *
+   * @return std::string the local cluster name, or "" if no local cluster was configured.
+   */
+  virtual const std::string localClusterName() const PURE;
```
I think `const std::string` doesn't do anything here, you probably either want `std::string` or `const std::string&`.

htuch commented on test/common/upstream/eds_test.cc:
```diff
@@ -346,6 +350,43 @@ TEST_F(EdsTest, EndpointHostsPerPriority) {
+  add_hosts_to_priority(1, 1);
+  bool initialized = false;
+  cluster_->initialize([&initialized] { initialized = true; });
+  EXPECT_THROW(cluster_->onConfigUpdate(resources), EnvoyException);
```
Probably prefer THROW_WITH_MESSAGE or add a EdsEnvoyException subclass here, so we know we're catching the right thing.

mattklein123 commented on source/common/upstream/eds.cc:
```diff
@@ -65,13 +65,13 @@ void EdsClusterImpl::onConfigUpdate(const ResourceVector& resources) {
   for (const auto& locality_lb_endpoint : cluster_load_assignment.endpoints()) {
     const uint32_t priority = locality_lb_endpoint.priority();
-    if (priority > 0 && info()->lbType() == LoadBalancerType::RingHash) {
+    if (false && priority > 0 && info()->lbType() == LoadBalancerType::RingHash) {
```
Did you mean for this false to be here?

mattklein123 commented on test/test_common/utility.h:
```diff
@@ -35,6 +35,12 @@ namespace Envoy {
+    statement;                                                                                     \
+  } catch (EnvoyException & e) {                                                                   \
+    ADD_FAILURE() << "Unexpected exception: " << std::string(e.what());                            \
+  }
```
super nit: I would add newline after this line.
